packages/loader/container-loader/src/container.ts

```diff
@@ -1642,6 +1686,19 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
         } else if (value === ConnectionState.Disconnected) {
             // Important as we process our own joinSession message through delta request
             this.pendingClientId = undefined;
+            // Only wait for "leave" message if we have some outstanding ops and the client was write client as
+            // server would not accept ops from read client.
+            if (this._deltaManager.clientSequenceNumber - this._deltaManager.clientSequenceNumberObserved > 0
```

Comment:
I do not think you need to check for "write" mode as it follows from the difference between clientSequenceNumber  & clientSequenceNumberObserved 

Given that deltaManager.ts has exactly same logic in connectCore(), can we instead extract it to a function on deltaManager and call from both places? It would be easier to find all such code when/if we redesign this area, otherwise it's impossible to find it.

Reply:
Actually we do need to check. Because client initially connects in "read" mode and then sends an op. It gets disconnected as it was in read mode. Now, in that the client no and observed one differs because of the op sent but we are never going to receive the leave message for the read client. So we don't need to wait in that case.

Reply:
Take a look at code and comment in connectCore() - we can't connect in "read" if above condition is true - DeltaManager will always try to connect as "write" if it knows there were some ops not acked.
Where you are correct is if we did connect as "write" and deltaManager reset these counters. If we lose that connection before seeing leave from previous one, then all the logic is broken (including this one, but also mostly because we reset this.prevClientLeftP above).

Reply:
So whenever we load a document we connect in Read mode and then client sends an op which moves client seq number and clientObserved seq number remains same. Read client gets disconnected. So in that when we will come here we will see the difference between the 2 numbers but because the client was read, then we don't need to wait for leave as we will not get it and also will not face any data corruption.